vta/apps/gemm/tests/python/chisel_accel.py

```diff
@@ -0,0 +1,159 @@
+
+    # create mask
+    slice_mask = 2**(slice_width)-1
+    # slice and pack
+    for x in range(len(A)):
+        for y in range(row):
+            C[y][x] = (np.uint64(A[x]) >> np.uint64(slice_width * y)) & np.uint64(slice_mask)
+    return C
+
+
+# A is a n*m matrix, B is a m*p matrix(not transposed yet)
+def matrix_multiply(A, B, w_width, a_width):
+    assert A.shape[1] == B.shape[0], "can't perform multiplication"
+    BT = B.transpose()
+    cycles = 0
+    C = np.zeros((A.shape[0], B.shape[1])).astype('uint64')
+    for i in range(A.shape[0]):
+        for j in range(B.shape[1]):
+            # C[i, j] = A[i].dot(BT[j])
+            A_sliced = slice(A[i], w_width)
+            B_sliced = slice(BT[j], a_width)
+
+            C[i, j] = compute(A_sliced, B_sliced, w_width, a_width)
+            test = test_accel(A_sliced, B_sliced, w_width, a_width)
+            cycles += test[1]
+            np.testing.assert_equal(C[i,j], A[i].astype('uint64').dot(BT[j]))
+            print("PASS SW serial & parallel")
+
+            np.testing.assert_equal(test[0], C[i, j])
+            print("PASS SW & HW bit serial")
+
+            np.testing.assert_equal(test[0], A[i].astype('uint64').dot(BT[j]))
+            print("PASS SW bit parallel & HW bit parallel")
+
+    print("result: ")
+    print(C)
+    print("ALL TESTS PASSED, cycles: " + str(cycles))
+    return C
+
+# takes 2 matrix input (sliced and packed)
+def compute(A, B, w_width, a_width):
+    assert A.shape[1] == B.shape[1], "sliced shape not match"
+    # reset hardware accumulator
+    accum = 0
+    for x in range(A.shape[0]):
+        for y in range(B.shape[0]):
+            # hardware implementation
+            accum += np.uint64(A[x]).dot(np.uint64(B[y])) << np.uint64(x*w_width + y*a_width)
+    # get value from accumulator
+    return accum
+
+def test_accel(A, B, w_width, a_width):
+    assert A.shape[1] == B.shape[1], "sliced shape not match"
+
+    dtype = A.dtype
+    ctx = tvm.cpu(0)
+    f = tsim.load_module()
+
+    a_arr = []
+    b_arr = []
+    for i in range(A.shape[0]):
+        list_a = np.zeros(A.shape[1]).astype(dtype)
+        for j in range(A.shape[1]):
+            list_a[j] = A[i][j]
+        a_arr.append(tvm.nd.array(list_a.astype(dtype), ctx))
+
+    for i in range(B.shape[0]):
+        list_b = np.zeros(B.shape[1]).astype(dtype)
+        for j in range(B.shape[1]):
+            list_b[j] = B[i][j]
+        b_arr.append(tvm.nd.array(list_b.astype(dtype), ctx))
+
+    cycles = 0
+
+    accum = tvm.nd.array(np.array([0]).astype("uint64"), ctx)
+    for i in range(len(a_arr)):
+        for j in range(len(b_arr)):
+            shift = np.uint8(i*w_width + j*a_width)
+            if i == 0 and j == 0: 
+                cycles += f(a_arr[i], b_arr[j], shift, accum, np.uint32(1)) # reset accumulator
+            else: 
+                cycles += f(a_arr[i], b_arr[j], shift, accum, np.uint32(0)) # no reset
+
+    return (accum.asnumpy()[0], cycles)
+
+# top_test
```

Comment:
It will be nice if we can match the python style for comments, see [here](https://github.com/dmlc/tvm/blob/master/vta/python/vta/environment.py#L80-L100)

Reply:
Fixed